cpp/distributed/dist-mnist.cpp

```diff
@@ -0,0 +1,192 @@
+      : conv1(torch::nn::Conv2dOptions(1, 10, 5)),
+        conv2(torch::nn::Conv2dOptions(10, 20, 5)),
+        fc1(320, 50),
+        fc2(50, 10) {
+    register_module("conv1", conv1);
+    register_module("conv2", conv2);
+    register_module("conv2_drop", conv2_drop);
+    register_module("fc1", fc1);
+    register_module("fc2", fc2);
+  }
+
+  torch::Tensor forward(torch::Tensor x) {
+    x = torch::relu(torch::max_pool2d(conv1->forward(x), 2));
+    x = torch::relu(
+        torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));
+    x = x.view({-1, 320});
+    x = torch::relu(fc1->forward(x));
+    x = torch::dropout(x, 0.5, is_training());
+    x = fc2->forward(x);
+    return torch::log_softmax(x, 1);
+  }
+
+  torch::nn::Conv2d conv1;
+  torch::nn::Conv2d conv2;
+  torch::nn::Dropout2d conv2_drop;
+  torch::nn::Linear fc1;
+  torch::nn::Linear fc2;
+};
+
+int main(int argc, char* argv[]) {
+  // MPI variables
+  int rank, numranks;
+  MPI_Init(&argc, &argv);
+  MPI_Comm_size(MPI_COMM_WORLD, &numranks);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  // end MPI variables
+
+  // Timer variables
+  auto tstart = 0.0;
+  auto tend = 0.0;
+
+  // TRAINING
+  // Read train dataset
+  const char* kDataRoot = "../data";
+  auto train_dataset =
+      torch::data::datasets::MNIST(kDataRoot)
+          .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
+          .map(torch::data::transforms::Stack<>());
+
+  // Distributed Random Sampler
+  auto data_sampler = torch::data::samplers::DistributedRandomSampler(
+      train_dataset.size().value(), numranks, rank, false);
+
+  auto num_train_samples_per_proc = train_dataset.size().value() / numranks;
+
+  // Generate dataloader
+  auto total_batch_size = 64;
+  auto batch_size_per_proc =
+      total_batch_size / numranks; // effective batch size in each processor
+  auto data_loader = torch::data::make_data_loader(
+      std::move(train_dataset), data_sampler, batch_size_per_proc);
+
+  // setting manual seed
+  torch::manual_seed(0);
+
+  auto model = std::make_shared<Model>();
+
+  auto learning_rate = 1e-2;
+
+  torch::optim::SGD optimizer(model->parameters(), learning_rate);
+
+  // Number of epochs
+  size_t num_epochs = 10;
+
+  // start timer
+  tstart = MPI_Wtime();
+
+  for (size_t epoch = 1; epoch <= num_epochs; ++epoch) {
+    size_t num_correct = 0;
+
+    for (auto& batch : *data_loader) {
+      auto ip = batch.data;
+      auto op = batch.target.squeeze();
+
+      // convert to required formats
+      ip = ip.to(torch::kF32);
+      op = op.to(torch::kLong);
+
+      // Reset gradients
+      model->zero_grad();
+
+      // Execute forward pass
+      auto prediction = model->forward(ip);
+
+      auto loss = torch::nll_loss(torch::log_softmax(prediction, 1), op);
+
+      // Backpropagation
+      loss.backward();
+
+      // Averaging the gradients of the parameters in all the processors
+      for (auto& param : model->named_parameters()) {
+        MPI_Allreduce(
```

Comment:
Can we create an MPI Process Group and call the process group `all_reduce` function instead?

Reply:
@osalpekar We can do that but will that help in some way?

Reply:
@soumyadipghosh good question - most of the reasons below exist for Python - let's check with @agolynski about ProcessGroup C++ API just to make sure this is allowed.

Using ProcessGroup is the preferred way of launching collectives and communication operations in PyTorch distributed. This way you can create a ProcessGroup using any underlying communication library you prefer (MPI, NCCL, Gloo, etc.), and simply call any of the collectives (allreduce, broadcast, etc.). Not only are you not tied to any comm library at the user program level, but you also gain some other benefits depending on the specific ProcessGroup type being used (error handling, timeout checking, perf optimizations, planned support for async Futures, etc.).

Reply:
@osalpekar Any update on how to proceed about the ProcessGroup C++ API?

Reply:
@soumyadipghosh Apologies for the delay! [Here's](https://github.com/pytorch/pytorch/blob/c115957df086fd795abaae18c6b9bc1e80928a18/torch/lib/c10d/ProcessGroupMPI.hpp) a link to the ProcessGroupMPI header file. You can simply construct the ProcessGroup and then call any collectives on that process group (which are implemented as class methods - for example, allreduce). If the MPI you are using is CUDA-aware, you can use CUDA tensors.

We also have a [NCCL ProcessGroup](https://github.com/pytorch/pytorch/blob/c115957df086fd795abaae18c6b9bc1e80928a18/torch/lib/c10d/ProcessGroupNCCL.hpp) that you can as above for using the NCCL backend. This way you will not have to call into the NCCL functions for initialization, creating communicators, etc. - it is all done for you, so you can simply create the ProcessGroup and then use the provided collectives (for example, `ProcessGroupNCCL::allreduce`). 

If you want to use both CPUs and GPUs while training, the best option would be Gloo (we also have a ProcessGroupGloo in the same folder as the above 2) and another option is to use MPI if your version of MPI is CUDA-aware.